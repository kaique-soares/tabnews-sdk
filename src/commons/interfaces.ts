export interface TabNewsApiError {
  name: string;
  message: string;
  action: string;
  statusCode: number;
  stack?: string;
  errorId?: string;
  requestId?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  context?: any;
  errorLocationCode?: string;
  key?: string;
  type?: string;
  databaseErrorCode?: string;
}
