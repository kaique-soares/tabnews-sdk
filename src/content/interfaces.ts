export type ContentStatus = 'published' | 'deleted' | 'draft';
export type CreateContentStatus = Exclude<ContentStatus, 'deleted'>;
export type ContentStrategy = 'new' | 'old' | 'relevant';

export interface ContentResponse {
  id: string;
  owner_id: string;
  owner_username: string;
  parent_id?: string;
  slug: string;
  title?: string;
  status: ContentStatus;
  source_url?: string;
  created_at: Date;
  updated_at: Date;
  published_at: Date;
  deleted_at?: Date;
  tabcoins: number;
  children_deep_count: number;
}

export interface GetContentListParams {
  username?: string;
  page?: number;
  per_page?: number;
  strategy?: ContentStrategy;
}

export interface ContentPagination {
  strategy: ContentStrategy;
  page: number;
  per_page: number;
  first_page: number;
  last_page: number;
  next_page?: number;
  previous_page?: number;
  total_rows: number;
}

export interface CreateContent {
  parent_id?: string;
  slug?: string;
  title?: string;
  body: string;
  status: CreateContentStatus;
  source_url?: string;
}

export interface GetContentParams {
  slug: string;
  username?: string;
}

export interface ContentDetailResponse extends ContentResponse {
  body: string;
}

export type UpdateContent = Partial<
  Omit<CreateContent, 'parent_id' | 'slug' | 'status'>
> & {
  slug: string;
  status?: ContentStatus;
  username?: string;
};

export interface ContentChildrenResponse extends ContentDetailResponse {
  children: ContentChildrenResponse[];
}

export interface ContentTabcoinsParams extends GetContentParams {
  transaction_type: 'credit' | 'debit';
}

export interface ContentTabcoinsResponse {
  tabcoins: number;
}
