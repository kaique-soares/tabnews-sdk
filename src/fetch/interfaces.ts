export type RequestConfig = Omit<RequestInit, 'headers' | 'body'> & {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  body?: any;
  path: string;
};
