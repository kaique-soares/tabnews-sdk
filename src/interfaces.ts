export interface TabNewsConfig {
  credentials?: {
    email?: string;
    password?: string;
  };
}
