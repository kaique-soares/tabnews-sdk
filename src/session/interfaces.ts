export interface SessionResponse {
  id: string;
  token: string;
  expires_at: Date;
  created_at: Date;
  updated_at: Date;
}
