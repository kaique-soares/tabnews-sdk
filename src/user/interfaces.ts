export interface UserResponse {
  id: string;
  username: string;
  email: string;
  description?: string;
  notifications: boolean;
  features: string[];
  tabcoins: number;
  tabcash: number;
  created_at: Date;
  updated_at: Date;
}
