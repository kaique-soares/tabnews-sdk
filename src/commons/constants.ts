export const TABNEWS_BASE_URL =
  process.env.TABNEWS_BASE_URL || 'https://www.tabnews.com.br/api/v1';

export const TABNEWS_ENDPOINTS = Object.freeze({
  session: '/sessions',
  user: '/user',
  content: '/contents',
});
