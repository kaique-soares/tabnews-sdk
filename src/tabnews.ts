import { TabNewsConfig } from './interfaces';

export class TabNews {
  constructor(readonly config: TabNewsConfig = {}) {
    if (!config.credentials) {
      config.credentials = {};
    }

    config.credentials.email =
      config.credentials.email || process.env.TABNEWS_CREDENTIALS_EMAIL;
    config.credentials.password =
      config.credentials.password || process.env.TABNEWS_CREDENTIALS_PASSWORD;

    this.config = config;
  }
}
