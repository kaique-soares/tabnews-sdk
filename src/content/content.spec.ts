import { expect, describe, it, afterEach, beforeEach } from 'vitest';

import { TABNEWS_ENDPOINTS, TABNEWS_HEADERS } from '@/commons';
import { TabNews } from '@/tabnews';
import {
  createTabNews,
  expectRequest,
  mockOnceApiError,
  mockOnceResponse,
  mockedRequest,
  mockedRequests,
  resetMocks,
} from '@test/utils';

import { GetContentParams } from './interfaces';

let tabNews: TabNews;

const linkHeader =
  '<https://www.tabnews.com.br/api/v1/contents?strategy=relevant&page=1&per_page=30>; rel="first", ' +
  '<https://www.tabnews.com.br/api/v1/contents?strategy=relevant&page=2&per_page=30>; rel="next", ' +
  '<https://www.tabnews.com.br/api/v1/contents?strategy=relevant&page=6&per_page=30>; rel="last"';

const lastPageLinkHeader =
  '<https://www.tabnews.com.br/api/v1/contents?strategy=relevant&page=1&per_page=30>; rel="first", ' +
  '<https://www.tabnews.com.br/api/v1/contents?strategy=relevant&page=5&per_page=30>; rel="prev", ' +
  '<https://www.tabnews.com.br/api/v1/contents?strategy=relevant&page=6&per_page=30>; rel="last"';

const content = {
  id: 'id',
  owner_id: 'owner_id',
  parent_id: null,
  slug: 'slug',
  title: 'title',
  status: 'published',
  source_url: 'https://source.url.com/source',
  created_at: '2023-09-19T12:16:04.812Z',
  updated_at: '2023-09-19T12:16:04.812Z',
  published_at: '2023-09-19T12:16:04.837Z',
  deleted_at: null,
  tabcoins: 1,
  owner_username: 'username',
  children_deep_count: 2,
};

describe('Content', () => {
  beforeEach(() => {
    tabNews = createTabNews();
  });

  afterEach(() => {
    resetMocks();
  });

  describe('get', () => {
    const mockContents = (link: string) => {
      mockOnceResponse(TABNEWS_ENDPOINTS.content, [content], {
        headers: {
          [TABNEWS_HEADERS.link]: link,
          [TABNEWS_HEADERS.paginationTotalRows]: '175',
        },
      });
    };

    it('should return all contents and pagination', async () => {
      mockContents(linkHeader);

      const response = await tabNews.content.getAll();

      expect(response.pagination).toMatchSnapshot();
      expect(response.contents).toMatchSnapshot();

      const request = mockedRequest();

      expectRequest(request).method.toBeGet();
    });

    it('should return correct page when is last page', async () => {
      mockContents(lastPageLinkHeader);

      const response = await tabNews.content.getAll();

      expect(response.pagination).toMatchSnapshot();

      expect(response.contents).toMatchSnapshot();

      const request = mockedRequest();

      expectRequest(request).method.toBeGet();
    });

    it('should send all content params correctly', () => {
      const paramsList: GetContentParams[] = [
        {
          page: 2,
          per_page: 10,
          strategy: 'new',
        },
        {
          per_page: 20,
          strategy: 'new',
        },
        {
          strategy: 'old',
        },
      ];

      paramsList.forEach(async (params) => {
        mockContents(linkHeader);

        await tabNews.content.getAll(params);
      });

      const [first, second, third] = mockedRequests();

      expectRequest(first).query('page').toBe('2');
      expectRequest(first).query('per_page').toBe('10');
      expectRequest(first).query('strategy').toBe('new');

      expectRequest(second).query('page').toBeNull();
      expectRequest(second).query('per_page').toBe('20');
      expectRequest(second).query('strategy').toBe('new');

      expectRequest(third).query('page').toBeNull();
      expectRequest(third).query('per_page').toBeNull();
      expectRequest(third).query('strategy').toBe('old');
    });

    it('should return error when parameter is invalid', () => {
      mockOnceApiError(TABNEWS_ENDPOINTS.content, {
        name: 'ValidationError',
        message: '"page" deve possuir um valor mínimo de 1.',
        action: 'Ajuste os dados enviados e tente novamente.',
        status_code: 400,
        error_id: 'd2ae3240-21db-45fe-9985-814aa317dfa1',
        request_id: '4b0a7574-9b89-41f8-8e52-97b972958a67',
        error_location_code: 'MODEL:VALIDATOR:FINAL_SCHEMA',
        key: 'page',
        type: 'number.min',
      });

      expect(() =>
        tabNews.content.getAll({
          page: 0,
        }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('create', () => {
    const mockCreateContent = () => {
      mockOnceResponse(TABNEWS_ENDPOINTS.content, {
        id: '502d02a1-2e38-4cc6-9e41-55c978e34c67',
        owner_id: '605144ec-9f0d-437b-96ee-1b5f4e54f5f9',
        parent_id: null,
        slug: 'e-opcional',
        title: 'test',
        body: 'test',
        status: 'draft',
        source_url: null,
        created_at: '2023-09-19T10:29:20.967Z',
        updated_at: '2023-09-19T10:29:20.967Z',
        published_at: null,
        deleted_at: null,
        owner_username: 'leoferreiralima',
        tabcoins: 0,
      });
    };

    it('should create content', async () => {
      mockCreateContent();

      const response = await tabNews.content.create({
        slug: 'e-opcional',
        title: 'test',
        body: 'test',
        status: 'draft',
        source_url: 'https://google.com',
      });

      expect(response).toMatchSnapshot();

      const request = mockedRequest();

      expectRequest(request).method.toBePost();
      expectRequest(request).body.toBe({
        slug: 'e-opcional',
        title: 'test',
        body: 'test',
        status: 'draft',
        source_url: 'https://google.com',
      });
    });

    it('should throw an error when create content with invalid parameters', () => {
      mockOnceApiError(TABNEWS_ENDPOINTS.content, {
        name: 'ValidationError',
        message: '"body" possui o valor inválido "null".',
        action: 'Ajuste os dados enviados e tente novamente.',
        status_code: 400,
        error_id: '68a48842-e7ec-4aa8-9549-eb69588cbfe2',
        request_id: '6d4278ca-f9e9-4187-a977-f5aab3f11200',
        error_location_code: 'MODEL:VALIDATOR:FINAL_SCHEMA',
        key: 'body',
        type: 'any.invalid',
      });

      expect(() =>
        tabNews.content.create({
          slug: 'e-opcional',
          title: 'test',
          body: 'invalid',
          status: 'draft',
          source_url: 'https://google.com',
        }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
