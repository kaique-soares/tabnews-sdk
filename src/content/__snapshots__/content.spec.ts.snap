// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Content > create > should create content 1`] = `
{
  "body": "test",
  "created_at": 2023-09-19T10:29:20.967Z,
  "deleted_at": null,
  "id": "502d02a1-2e38-4cc6-9e41-55c978e34c67",
  "owner_id": "605144ec-9f0d-437b-96ee-1b5f4e54f5f9",
  "owner_username": "leoferreiralima",
  "parent_id": null,
  "published_at": null,
  "slug": "e-opcional",
  "source_url": null,
  "status": "draft",
  "tabcoins": 0,
  "title": "test",
  "updated_at": 2023-09-19T10:29:20.967Z,
}
`;

exports[`Content > create > should throw an error when create content with invalid parameters 1`] = `"\\"body\\" possui o valor inválido \\"null\\"."`;

exports[`Content > get > should get all contents and pagination 1`] = `
{
  "first_page": 1,
  "last_page": 6,
  "next_page": 2,
  "page": 1,
  "per_page": 30,
  "previous_page": undefined,
  "strategy": "relevant",
  "total_rows": 175,
}
`;

exports[`Content > get > should get all contents and pagination 2`] = `
[
  {
    "children_deep_count": 2,
    "created_at": 2023-09-19T12:16:04.812Z,
    "deleted_at": null,
    "id": "id",
    "owner_id": "owner_id",
    "owner_username": "username",
    "parent_id": null,
    "published_at": 2023-09-19T12:16:04.837Z,
    "slug": "slug",
    "source_url": "https://source.url.com/source",
    "status": "published",
    "tabcoins": 1,
    "title": "title",
    "updated_at": 2023-09-19T12:16:04.812Z,
  },
]
`;

exports[`Content > get > should get all contents for a specific user 1`] = `
{
  "first_page": 1,
  "last_page": 6,
  "next_page": 2,
  "page": 1,
  "per_page": 30,
  "previous_page": undefined,
  "strategy": "relevant",
  "total_rows": 175,
}
`;

exports[`Content > get > should get all contents for a specific user 2`] = `
[
  {
    "children_deep_count": 2,
    "created_at": 2023-09-19T12:16:04.812Z,
    "deleted_at": null,
    "id": "id",
    "owner_id": "owner_id",
    "owner_username": "username",
    "parent_id": null,
    "published_at": 2023-09-19T12:16:04.837Z,
    "slug": "slug",
    "source_url": "https://source.url.com/source",
    "status": "published",
    "tabcoins": 1,
    "title": "title",
    "updated_at": 2023-09-19T12:16:04.812Z,
  },
]
`;

exports[`Content > get > should get all contents for current user 1`] = `
{
  "first_page": 1,
  "last_page": 6,
  "next_page": 2,
  "page": 1,
  "per_page": 30,
  "previous_page": undefined,
  "strategy": "relevant",
  "total_rows": 175,
}
`;

exports[`Content > get > should get all contents for current user 2`] = `
[
  {
    "children_deep_count": 2,
    "created_at": 2023-09-19T12:16:04.812Z,
    "deleted_at": null,
    "id": "id",
    "owner_id": "owner_id",
    "owner_username": "username",
    "parent_id": null,
    "published_at": 2023-09-19T12:16:04.837Z,
    "slug": "slug",
    "source_url": "https://source.url.com/source",
    "status": "published",
    "tabcoins": 1,
    "title": "title",
    "updated_at": 2023-09-19T12:16:04.812Z,
  },
]
`;

exports[`Content > get > should get content by slug 1`] = `
{
  "body": "body",
  "children_deep_count": 2,
  "created_at": 2023-09-19T12:16:04.812Z,
  "deleted_at": null,
  "id": "id",
  "owner_id": "owner_id",
  "owner_username": "username",
  "parent_id": null,
  "published_at": 2023-09-19T12:16:04.837Z,
  "slug": "slug",
  "source_url": "https://source.url.com/source",
  "status": "published",
  "tabcoins": 1,
  "title": "title",
  "updated_at": 2023-09-19T12:16:04.812Z,
}
`;

exports[`Content > get > should get content by slug for current user 1`] = `
{
  "body": "body",
  "children_deep_count": 2,
  "created_at": 2023-09-19T12:16:04.812Z,
  "deleted_at": null,
  "id": "id",
  "owner_id": "owner_id",
  "owner_username": "username",
  "parent_id": null,
  "published_at": 2023-09-19T12:16:04.837Z,
  "slug": "slug",
  "source_url": "https://source.url.com/source",
  "status": "published",
  "tabcoins": 1,
  "title": "title",
  "updated_at": 2023-09-19T12:16:04.812Z,
}
`;

exports[`Content > get > should return correct page when is last page 1`] = `
{
  "first_page": 1,
  "last_page": 6,
  "next_page": undefined,
  "page": 6,
  "per_page": 30,
  "previous_page": 5,
  "strategy": "relevant",
  "total_rows": 175,
}
`;

exports[`Content > get > should return correct page when is last page 2`] = `
[
  {
    "children_deep_count": 2,
    "created_at": 2023-09-19T12:16:04.812Z,
    "deleted_at": null,
    "id": "id",
    "owner_id": "owner_id",
    "owner_username": "username",
    "parent_id": null,
    "published_at": 2023-09-19T12:16:04.837Z,
    "slug": "slug",
    "source_url": "https://source.url.com/source",
    "status": "published",
    "tabcoins": 1,
    "title": "title",
    "updated_at": 2023-09-19T12:16:04.812Z,
  },
]
`;

exports[`Content > get > should throw an error when content not found 1`] = `"O conteúdo informado não foi encontrado no sistema."`;

exports[`Content > get > should throw an error when parameter is invalid 1`] = `"\\"page\\" deve possuir um valor mínimo de 1."`;
